Index: obuspa-11.0.3/src/core/usp_log.c
===================================================================
--- obuspa-11.0.3.orig/src/core/usp_log.c
+++ obuspa-11.0.3/src/core/usp_log.c
@@ -117,6 +117,7 @@ int USP_LOG_SetFile(const char *file)
     if ((file == NULL) || strcmp(file, "syslog")==0)
     {
         log_fd = NULL;
+        openlog("obuspa", LOG_PID | LOG_NDELAY, LOG_DAEMON);
         return USP_ERR_OK;
     }
 
